ublox/src/ubx_packets/packets.rs

```diff
@@ -1939,111 +1941,123 @@ struct RxmRtcm {
-        from = EsfMeasInfo::new,
-        flatten,
-        flat_fields = ["flags", "u16", "id", "u16", "data", "U32Iter"]
+        map_type = Option<u32>,
+        from = EsfMeas::calib_tag,
+        size_fn = calib_tag_len,
     )]
-    //, "calib_tag", "Option<u32>"
-    info: [u8; 0],
+    calib_tag: [u8; 0],
 }
 
-#[derive(serde::Serialize, Debug)]
-pub struct EsfMeasInfo<'a> {
-    pub flags: u16,
-    pub id: u16,
-    pub data: U32Iter<'a>,
-    pub calib_tag: Option<u32>,
+impl EsfMeas {
+    fn calib_tag(bytes: &[u8]) -> Option<u32> {
+        bytes.try_into().ok().map(u32::from_le_bytes)
+    }
 }
 
-#[ubx_iter]
-pub struct U32Iter<'a>(core::slice::ChunksExact<'a, u8>);
+impl<'a> EsfMeasRef<'a> {
+    fn data_len(&self) -> usize {
+        ((self.flags() >> 11 & 0x1f) as usize) * 4
+    }
+
+    fn calib_tag_len(&self) -> usize {
+        if self.flags() & 0x8 != 0 {
+            4
+        } else {
+            0
+        }
+    }
+}
 
-impl<'a> U32Iter<'a> {
+#[derive(Debug, serde::Serialize)]
+pub struct EsfMeasData {
+    pub data_type: u8,
+    pub data_field: u32,
+}
+
+#[derive(Debug, Clone)]
+pub struct EsfMeasDataIter<'a>(core::slice::ChunksExact<'a, u8>);
+
+impl<'a> EsfMeasDataIter<'a> {
     fn new(bytes: &'a [u8]) -> Self {
-        U32Iter(bytes.chunks_exact(4))
+        Self(bytes.chunks_exact(4))
     }
 
     fn is_valid(bytes: &'a [u8]) -> bool {
         bytes.len() % 4 == 0
     }
 }
 
-impl<'a> core::iter::Iterator for U32Iter<'a> {
-    type Item = u32;
+impl<'a> core::iter::Iterator for EsfMeasDataIter<'a> {
+    type Item = EsfMeasData;
 
     fn next(&mut self) -> Option<Self::Item> {
-        self.0
-            .next()
-            .map(|bytes| u32::from_le_bytes(bytes.try_into().unwrap()))
-    }
-}
-
-impl<'a> EsfMeasInfo<'a> {
-    fn new(bytes: &'a [u8]) -> Self {
-        let flags = u16::from_le_bytes(bytes[0..2].try_into().unwrap());
-        let id = u16::from_le_bytes(bytes[2..4].try_into().unwrap());
-        let num_meas = usize::from(flags >> 11 & 0x1F);
-        let data = U32Iter::new(&bytes[4..num_meas * 4]);
-        let calib_tag = if flags & 0x8 != 0 {
-            let slice = bytes[num_meas * 4 + 4..num_meas * 4 + 8]
-                .try_into()
-                .unwrap();
-            Some(u32::from_le_bytes(slice))
-        } else {
-            None
-        };
-
-        Self {
-            flags,
-            id,
-            data,
-            calib_tag,
-        }
+        let data = self.0.next()?.try_into().map(u32::from_le_bytes).unwrap();
+        Some(EsfMeasData {
+            data_type: ((data & 0x3F000000) >> 24).try_into().unwrap(),
+            data_field: data & 0xFFFFFF,
+        })
     }
 }
 
 #[ubx_packet_recv]
 #[ubx(class = 0x10, id = 0x03, max_payload_len = 1240)]
 struct EsfRaw {
-    msss: u32,
-    #[ubx(map_type = EsfRawIter,
-    from = EsfRawIter::new,
-    is_valid = EsfRawIter::is_valid)]
-    iter: [u8; 0],
+    reserved1: u32,
```

Comment:
would this still be msss or reserved1?

Reply:
I think it's called reserve1 according this the docs? https://content.u-blox.com/sites/default/files/products/documents/u-blox8-M8_ReceiverDescrProtSpec_UBX-13003221.pdf#%5B%7B%22num%22%3A591%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C0%2C841.89%2Cnull%5D

I guess it should be a [u8; 4] tho

Reply:
https://github.com/swift-nav/gnss-converters-private/blob/8fd518e2a94b6c28f390e8207676889aa40f1f64/c/libubx/src/decode.c#L614

i think we called it msss in gnss-conv

Reply:
Ohh we can do that then